rolling-shutter/collator/batchhandler/sequencer/statecache.go

```diff
@@ -27,34 +30,104 @@ type EthClient interface {
+	return &KeyedMutex[T]{mu: sync.RWMutex{}, muxs: make(map[T]*countedMux)}
+}
+
+type KeyedMutex[T comparable] struct {
+	mu   sync.RWMutex
+	muxs map[T]*countedMux
+}
+
+func (ml *KeyedMutex[T]) call(method string, v T, delta int) {
+	ml.mu.Lock()
+	vmu, ok := ml.muxs[v]
+	if !ok {
+		vmu = &countedMux{
+			mux: &sync.RWMutex{},
+			cnt: 0,
+		}
+		ml.muxs[v] = vmu
+	}
+	var addUint32 uint32
+	if delta < 0 {
+		// recommended operation to substract from counter
+		addUint32 = ^uint32((-delta) - 1)
+	} else {
+		addUint32 = uint32(delta)
+	}
+	atomic.AddUint32(&vmu.cnt, addUint32)
+	ml.mu.Unlock()
+	// e.g. for method="Lock", this translates to:
+	// vmu.mux.Lock()
+	reflect.ValueOf(vmu.mux).MethodByName(method).Call([]reflect.Value{})
+	ml.mu.Lock()
+	if vmu.cnt <= 0 {
+		delete(ml.muxs, v)
+	}
+	ml.mu.Unlock()
+}
+
+func (ml *KeyedMutex[T]) Lock(v T) {
+	ml.call("Lock", v, 1)
+}
+
+func (ml *KeyedMutex[T]) Unlock(v T) {
+	ml.call("Unlock", v, -1)
+}
+
+func (ml *KeyedMutex[T]) RLock(v T) {
+	ml.call("RLock", v, 1)
+}
+
+func (ml *KeyedMutex[T]) RUnlock(v T) {
+	ml.call("RUnlock", v, -1)
+}
+
+func NewCached(client EthClient, atBlockNumber *big.Int) *Cached {
+	return &Cached{
+		balances:      make(map[common.Address]*big.Int),
+		nonces:        make(map[common.Address]uint64),
+		balancesLocks: NewKeyedMutex[common.Address](),
+		noncesLocks:   NewKeyedMutex[common.Address](),
 		Client:        client,
-		AtBlockNumber: nil,
-		balances:      make(map[common.Address]*big.Int, 0),
-		nonces:        make(map[common.Address]uint64, 0),
+		AtBlockNumber: atBlockNumber,
 	}
 }
 
-// ChainBatchCache tracks the state of account's nonces and balances
+// Cached tracks the state of account's nonces and balances
 // for a certain Batch.
 // If an address is not cached yet, it polls initial balances or nonces
 // on a GetBalance or GetNonce call from the underlying ethereum node via the
-// ChainBatchCache.Client. Then the value is cached and never polled again for
+// Cached.Client. Then the value is cached and never polled again for
 // that address.
 // This allows to poll chain-state and then modify it locally, e.g. while
 // accepting user transactions to be proposed as the next block to the sequencer.
-type ChainBatchCache struct {
-	balances map[common.Address]*big.Int
-	nonces   map[common.Address]uint64
+type Cached struct {
+	balances      map[common.Address]*big.Int
+	nonces        map[common.Address]uint64
+	balancesLocks *KeyedMutex[common.Address]
+	noncesLocks   *KeyedMutex[common.Address]
 
 	Client        EthClient
 	AtBlockNumber *big.Int
 }
 
 // GetBalance polls and caches the state of account `a` balance at the
 // block number ChainBatchCache.AtBlockNumber.
-func (c *ChainBatchCache) GetBalance(ctx context.Context, a common.Address) (*big.Int, error) {
+func (c *Cached) GetBalance(ctx context.Context, a common.Address) (*big.Int, error) {
+	// write lock because we eventually write to the dict
```

Comment:
why not make `getBalance` do this? Same for `subBalance` etc.

Reply:
This doesn't work, because subBalance calls `getBalance`.
If the method in the lowest layer acquires the mutex you would get nested write-mutex calls, which will deadlock